Web Inspector: Backend fails to bind WebAnimation with custom properties in keyframes

patrickangle opened the pull request:
#### 8a7cdf98234b35d9d4e5a1b2eb6daf449bef8258
<pre>
Web Inspector: Backend fails to bind WebAnimation with custom properties in keyframes
<a href="https://bugs.webkit.org/show_bug.cgi?id=251173">https://bugs.webkit.org/show_bug.cgi?id=251173</a>
rdar://104555265

Reviewed by Antti Koivisto.

InspectorAnimationAgent::didCreateAnimation is called during WebCore::Document::updateStyleIfNeeded, but in order to
resolve custom properties of animations, we may have to again enter WebCore::Document::updateStyleIfNeeded via
WebCore::ComputedStyleExtractor::updateStyleIfNeededForProperty. Instead, we should wait until after styles have been
updated which will prevent us from reentering WebCore::Document::updateStyleIfNeeded in order to resolve custom
properties.

This also has the benefit that the frontend will no longer receive a WebAnimation immediately followed by an update to
its name, since the name is set later. This allows us to capture the animation after it has been fully set up via the
updating of styles.

* LayoutTests/inspector/animation/lifecycle-web-animation-expected.txt:
* LayoutTests/inspector/animation/lifecycle-web-animation.html:
- Add test case to exercise the creation/deletion of animations with custom properties.

* LayoutTests/inspector/animation/resources/lifecycle-utilities.js:
- As a result of defering animation binding until later, the name no longer changes immediately after binding.

* Source/WebCore/inspector/agents/InspectorAnimationAgent.cpp:
(WebCore::InspectorAnimationAgent::InspectorAnimationAgent):
(WebCore::InspectorAnimationAgent::enable):
(WebCore::InspectorAnimationAgent::didCreateWebAnimation):
(WebCore::InspectorAnimationAgent::animationBindingTimerFired):
(WebCore::InspectorAnimationAgent::bindAnimation):
(WebCore::InspectorAnimationAgent::reset):
* Source/WebCore/inspector/agents/InspectorAnimationAgent.h:

Canonical link: <a href="https://commits.webkit.org/259446@main">https://commits.webkit.org/259446@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/85ee31f9f3a9bc1c02c49b1ee2ab0a011d96dc46

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104819 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13899 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37717 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114089 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/174288 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108727 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15027 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4824 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97146 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/113045 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110579 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/11614 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94618 "Found 3 new API test failures: TestWebKitAPI.WebAuthenticationPanel.PublicKeyCredentialRequestOptionsMaximum, TestWebKitAPI.WebAuthenticationPanel.PublicKeyCredentialCreationOptionsMaximum1, TestWebKitAPI.WebAuthenticationPanel.PublicKeyCredentialCreationOptionsMaximum2 (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39135 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93474 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26232 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/80799 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7246 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27592 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7345 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4194 "Found 2 new test failures: fast/images/avif-as-image.html, fast/images/avif-heif-container-as-image.html (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13397 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47142 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6513 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9131 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/inspector/agents/InspectorAnimationAgent.cpp:
```diff
@@ -229,6 +229,7 @@ InspectorAnimationAgent::InspectorAnimationAgent(PageAgentContext& context)
     , m_backendDispatcher(Inspector::AnimationBackendDispatcher::create(context.backendDispatcher, this))
     , m_injectedScriptManager(context.injectedScriptManager)
     , m_inspectedPage(context.inspectedPage)
+    , m_animationBindingTimer(*this, &InspectorAnimationAgent::animationBindingTimerFired)
```
I suppose an alternative to the timer would be to do the work in `InspectorInstrumentation::didRecalculateStyle`. But maybe the timer has other benefits too (like allowing things to coalesce).

patrickangle replied:
A benefit the approach in this patch is providing us is not having to separately handle WebAnimations created during `updateStyleIfNeeded` and those created via JavaScript directly, since those create by JavaScript aren't guaranteed to be followed by a `didRecalculateStyle` as far as I can tell.

patrickangle committed:
Web Inspector: Backend fails to bind WebAnimation with custom properties in keyframes
https://bugs.webkit.org/show_bug.cgi?id=251173
rdar://104555265

Reviewed by Antti Koivisto.

InspectorAnimationAgent::didCreateAnimation is called during WebCore::Document::updateStyleIfNeeded, but in order to
resolve custom properties of animations, we may have to again enter WebCore::Document::updateStyleIfNeeded via
WebCore::ComputedStyleExtractor::updateStyleIfNeededForProperty. Instead, we should wait until after styles have been
updated which will prevent us from reentering WebCore::Document::updateStyleIfNeeded in order to resolve custom
properties.

This also has the benefit that the frontend will no longer receive a WebAnimation immediately followed by an update to
its name, since the name is set later. This allows us to capture the animation after it has been fully set up via the
updating of styles.

* LayoutTests/inspector/animation/lifecycle-web-animation-expected.txt:
* LayoutTests/inspector/animation/lifecycle-web-animation.html:
- Add test case to exercise the creation/deletion of animations with custom properties.

* LayoutTests/inspector/animation/resources/lifecycle-utilities.js:
- As a result of defering animation binding until later, the name no longer changes immediately after binding.

* Source/WebCore/inspector/agents/InspectorAnimationAgent.cpp:
(WebCore::InspectorAnimationAgent::InspectorAnimationAgent):
(WebCore::InspectorAnimationAgent::enable):
(WebCore::InspectorAnimationAgent::didCreateWebAnimation):
(WebCore::InspectorAnimationAgent::animationBindingTimerFired):
(WebCore::InspectorAnimationAgent::bindAnimation):
(WebCore::InspectorAnimationAgent::reset):
* Source/WebCore/inspector/agents/InspectorAnimationAgent.h:

Canonical link: https://commits.webkit.org/259446@main
